File: src/connectionView.js
import GObject from 'gi://GObject'
import Gtk from 'gi://Gtk'
import Adw from 'gi://Adw'
import Tlg from 'gi://Tlg'

export class ConnectionView extends Adw.Bin {
    static {
        GObject.registerClass({
            GTypeName: 'TrilogyConnectionView',
            Properties: {
                'connection': GObject.ParamSpec.object(
                    'connection',
                    'Connection',
                    'The connection to view',
                    GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
                    Tlg.Connection
                ),
                'show-start-title-buttons': GObject.ParamSpec.boolean(
                    'show-start-title-buttons',
                    'Show Start Title Buttons',
                    'Whether the headerbar should show the start title buttons',
                    GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
                    false
                )
            },
            Template: 'resource:///com/rutins/Trilogy/connectionView.ui'
        }, this)
    }

    get connection() {
        return this._connection ?? Tlg.Connection.new('No Connection Selected', '');
    }

    set connection(conn) {
        if(this._connection == conn) return;
        this._connection = conn;
        this.notify('connection');
        this.refreshConnection();
    }

    get show_start_title_buttons() {
        return this._show_start_title_buttons ?? false;
    }

    set show_start_title_buttons(sstb) {
        if(this._show_start_title_buttons == sstb) return;
        this._show_start_title_buttons = sstb;
        this.notify('show-start-title-buttons');
    }

    refreshConnection() {
        const conn = this.connection;

        if(!conn.connected) {
            conn.ensure();
        }
    }
}

File: src/connection_view.py
from gi.repository import Gtk, GObject, Gio, Adw, Tlg

@Gtk.Template(resource_path='/com/rutins/Trilogy/connectionView.ui')
class ConnectionView(Adw.Bin):
    __gtype_name__ = 'TrilogyConnectionView'
    __gsignals__ = {
        'go-to-navigation': (GObject.SIGNAL_RUN_LAST, GObject.TYPE_NONE, ()),
        'open-settings': (GObject.SIGNAL_RUN_LAST, GObject.TYPE_NONE, (str,))
    }

    _overview_page = None

    tab_view = Gtk.Template.Child()

    connection = GObject.Property(type=Tlg.Connection)
    show_start_title_buttons = GObject.Property(type=bool, default=False)

    def __init__(self, conn):
        super().__init__()
        self.props.connection = conn
        self.refresh_connection()

    @Gtk.Template.Callback()
    def _back_clicked(self, *args):
        self.emit('go-to-navigation')
    
    @Gtk.Template.Callback()
    def _open_settings(self, *args):
        self.emit('open-settings', self.props.connection.props.name)

    def refresh_connection(self):
        conn = self.props.connection

        if not conn.props.connected:
            conn.ensure()
            if self._overview_page is not None:
                page = self.tab_view.get_page(self._overview_page)
                self.tab_view.set_page_pinned(page, False)
                self.tab_view.close_page(page)
            self._overview_page = conn.props.db.overview_page()
            self.tab_view.append_pinned(self._overview_page)
            page = self.tab_view.get_page(self._overview_page)
            page.set_icon(Gio.Icon.new_for_string('about-symbolic'))
            page.set_title('Overview')